### PR TITLE
feat: Add episode replay buffer for RL agents

### PR DESCRIPTION
**Pull request by joshgreaves**
# User description
## Summary

Adds a production-ready episode replay buffer implementation for reinforcement learning agents with full asyncio support.

### Key Features

- **Episode-based storage**: Store complete episodes with explicit lifecycle control (start, append transitions, end)
- **N-step sampling**: Sample n-step experiences uniformly across all time steps with automatic episode boundary handling
- **Asyncio-safe**: All operations use internal locking for safe concurrent access by multiple tasks
- **Capacity management**: Automatic eviction of oldest episodes when capacity limits (max_episodes or max_steps) are exceeded
- **Memory-efficient**: Avoids state duplication by storing observations sequentially and deriving next_obs during sampling

### Implementation Details

- Storage format: `observations[t]`, `actions[t]`, `rewards[t]` per episode
- Uniform sampling over all valid time steps (not episodes)
- N-step windows never cross episode boundaries
- Eviction policy: oldest finished episodes first, then oldest in-progress
- Thread-safety through asyncio.Lock (designed for asyncio only, not threading.Thread)

### Test Coverage

Comprehensive test suite (35 tests) covering:
- Episode lifecycle operations
- N-step sampling with boundary conditions
- Concurrent access patterns
- Capacity and eviction policies
- Edge cases and error conditions

All tests pass, ruff checks pass (lint + format).

### Files Changed

- `src/ares/contrib/rl/replay_buffer.py`: Core implementation (585 lines)
- `tests/contrib/rl/test_replay_buffer.py`: Test suite (662 lines)
- `pyproject.toml`: Add pytest-asyncio dependency and config

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

# Generated description

Below is a concise technical summary of the changes proposed in this PR:
```mermaid
graph LR
EpisodeReplayBuffer_("EpisodeReplayBuffer"):::added
Episode_("Episode"):::added
ReplaySample_("ReplaySample"):::added
compute_discounted_return_("compute_discounted_return"):::added
EpisodeReplayBuffer_ -- "Added buffer storing Episode instances with capacity eviction" --> Episode_
EpisodeReplayBuffer_ -- "Adds n-step ReplaySample construction for uniform sampling" --> ReplaySample_
ReplaySample_ -- "ReplaySample.reward computes discounted return via compute_discounted_return" --> compute_discounted_return_
classDef added stroke:#15AA7A
classDef removed stroke:#CD5270
classDef modified stroke:#EDAC4C
linkStyle default stroke:#CBD5E1,font-size:13px
```

Introduces a production-ready <code>EpisodeReplayBuffer</code> within the <code>ares.contrib.rl</code> module, providing an asyncio-compatible mechanism for episodic experience storage, uniform n-step sampling, and automatic capacity management for reinforcement learning agents. Establishes the necessary package structure and includes a comprehensive test suite to validate the buffer's lifecycle, sampling, and eviction policies.
<table><tr><th>Topic</th><th>Details</th><tr><td><a href=https://baz.co/changes/withmartian/ares/19?tool=ast&topic=Testing+%26+Project+Setup>Testing & Project Setup</a>
        </td><td>Adds a comprehensive test suite for the <code>EpisodeReplayBuffer</code> covering lifecycle, sampling, concurrency, and eviction, and updates <code>pyproject.toml</code> to include <code>pytest-asyncio</code> and adjust test path configurations.<details><summary>Modified files (4)</summary><ul><li>pyproject.toml</li>
<li>src/ares/contrib/rl/replay_buffer_test.py</li>
<li>tests/contrib/__init__.py</li>
<li>tests/contrib/rl/__init__.py</li></ul></details><details><summary>Latest Contributors(2)</summary><table><tr><th>User</th><th>Commit</th><th>Date</th></tr><tr><td>joshua.greaves@gmail.com</td><td>fix-avoid-os.getlogin-...</td><td>January 13, 2026</td></tr>
<tr><td>ryanscais3@gmail.com</td><td>Add-DM-Env-Interface-3</td><td>December 18, 2025</td></tr></table></details></td></tr>
<tr><td><a href=https://baz.co/changes/withmartian/ares/19?tool=ast&topic=Episodic+Replay+Buffer>Episodic Replay Buffer</a>
        </td><td>Implements the <code>EpisodeReplayBuffer</code> class and its associated data structures (<code>Episode</code>, <code>ReplaySample</code>) within the <code>ares.contrib.rl</code> package, enabling efficient, asyncio-compatible storage and n-step sampling of agent experiences with capacity management.<details><summary>Modified files (3)</summary><ul><li>src/ares/contrib/__init__.py</li>
<li>src/ares/contrib/rl/__init__.py</li>
<li>src/ares/contrib/rl/replay_buffer.py</li></ul></details><details><summary>Latest Contributors(0)</summary><table><tr><th>User</th><th>Commit</th><th>Date</th></tr></table></details></td></tr></table>
This pull request is reviewed by Baz. Review like a pro on <a href=https://baz.co/changes/withmartian/ares/19?tool=ast>(Baz)</a>.

### CODE REVIEW

File: src/ares/contrib/rl/replay_buffer.py [hunk 1]
```diff
@@ -0,0 +1,585 @@
...
+        observations: List of observations in temporal order
+        actions: List of actions taken
+        rewards: List of rewards received
+        status: Current status of the episode
+        start_time: Timestamp when episode started (for eviction policy)
+    """
+
+    episode_id: str
+    agent_id: str
+    observations: list[Any] = field(default_factory=list)
```
[the remark this answers is not in the document]
Comment (rowan-assistant) [reply src/ares/contrib/rl/replay_buffer.py]: ✅ Already using PEP 695 generics - Episode[ObservationType, ActionType] and ReplaySample[ObservationType, ActionType]

File: pyproject.toml [hunk 2]
```diff
@@ -49,3 +50,5 @@ examples = [
 [tool.pytest.ini_options]
 testpaths = ["src"]
 python_files = ["test_*.py", "*_test.py"]
+asyncio_mode = "auto"
+asyncio_default_fixture_loop_scope = "function"
```
[the remark this answers is not in the document]
Comment (rowan-assistant) [reply pyproject.toml]: ✅ Fixed in acaa54c

File: src/ares/contrib/rl/replay_buffer.py [hunk 3]
```diff
@@ -0,0 +1,585 @@
...
+            next_obs = episode.observations[end_idx]
+        else:
+            # Episode ended; last observation should be at index end_idx-1+1 = end_idx
+            # But if observations has length num_steps+1, then end_idx could equal num_steps
+            # In that case, the last observation is observations[num_steps]
+            # Let's ensure observations has the final obs
+            if len(episode.observations) > end_idx:
+                next_obs = episode.observations[end_idx]
+            else:
+                # Fallback: use the last available observation
```
[the remark this answers is not in the document]
Comment (rowan-assistant) [reply src/ares/contrib/rl/replay_buffer.py]: ✅ Fixed in 61be6d0

File: src/ares/contrib/rl/replay_buffer.py [hunk 4]
```diff
@@ -0,0 +1,585 @@
...
+    agent_id: str
+    observations: list[Any] = field(default_factory=list)
+    actions: list[Any] = field(default_factory=list)
+    rewards: list[float] = field(default_factory=list)
+    status: EpisodeStatus = EpisodeStatus.IN_PROGRESS
+    start_time: float = field(default_factory=time.time)
+
+    def __len__(self) -> int:
+        """Return the number of valid (obs, action, reward) tuples (i.e., len(actions))."""
+        return len(self.actions)
```
[the remark this answers is not in the document]
Comment (rowan-assistant) [reply src/ares/contrib/rl/replay_buffer.py]: ✅ Fixed in acaa54c

File: src/ares/contrib/rl/replay_buffer.py [hunk 5]
```diff
@@ -0,0 +1,585 @@
...
+
+    IN_PROGRESS = "in_progress"
+    TERMINAL = "terminal"  # Episode ended naturally (e.g., goal reached, death)
+    TRUNCATED = "truncated"  # Episode ended due to time limit or external constraint
+
+
+@dataclass
+class Episode:
+    """
+    An episode containing sequences of observations, actions, and rewards.
```
[the remark this answers is not in the document]
Comment (rowan-assistant) [reply src/ares/contrib/rl/replay_buffer.py]: Fixed in acaa54c - Updated Episode docstring to single-line format.

File: src/ares/contrib/rl/replay_buffer.py [hunk 6]
```diff
@@ -0,0 +1,585 @@
...
+
+    def __len__(self) -> int:
+        """Return the number of valid (obs, action, reward) tuples (i.e., len(actions))."""
+        return len(self.actions)
+
+
+@dataclass
+class NStepSample:
+    """
+    A sampled n-step experience for training.
```
[the remark this answers is not in the document]
Comment (rowan-assistant) [reply src/ares/contrib/rl/replay_buffer.py]: Fixed in acaa54c - Updated ReplaySample docstring to single-line format.

File: src/ares/contrib/rl/replay_buffer.py [hunk 7]
```diff
@@ -0,0 +1,585 @@
...
+        discount_powers: [gamma^0, gamma^1, ..., gamma^{m-1}] for computing returns
+        start_step: The starting step index t
+        actual_n: The actual number of steps m (may be < n if episode ends)
+        gamma: The discount factor used
+    """
+
+    episode_id: str
+    agent_id: str
+    obs_t: Any
+    action_t: Any
```
[the remark this answers is not in the document]
Comment (rowan-assistant) [reply src/ares/contrib/rl/replay_buffer.py]: Fixed in acaa54c - Added generic types: Episode[ObservationType, ActionType] and ReplaySample[ObservationType, ActionType] using PEP 695 syntax.

File: src/ares/contrib/rl/replay_buffer.py [hunk 8]
```diff
@@ -0,0 +1,585 @@
...
+    terminal: bool
+    discount_powers: list[float]
+    start_step: int
+    actual_n: int
+    gamma: float
+
+
+def compute_discounted_return(rewards: list[float], gamma: float) -> float:
+    """
+    Compute the discounted return from a sequence of rewards.
```
[the remark this answers is not in the document]
Comment (rowan-assistant) [reply src/ares/contrib/rl/replay_buffer.py]: Fixed in acaa54c - Updated compute_discounted_return docstring to single-line format.

File: src/ares/contrib/rl/replay_buffer.py [hunk 9]
```diff
@@ -0,0 +1,585 @@
...
+        # Track episodes by agent for potential future use
+        self._agent_episodes: dict[str, list[str]] = defaultdict(list)
+
+    async def start_episode(
+        self,
+        agent_id: str,
+        episode_id: str | None = None,
+    ) -> str:
+        """
+        Start a new episode.
```
[the remark this answers is not in the document]
Comment (rowan-assistant) [reply src/ares/contrib/rl/replay_buffer.py]: Fixed in acaa54c - Updated all docstrings to have first line on same line as opening """. Applied throughout the file.

File: src/ares/contrib/rl/replay_buffer_test.py
[the remark this answers is not in the document]
Comment (rowan-assistant) [reply src/ares/contrib/rl/replay_buffer_test.py]: Fixed in acaa54c - Moved test file from tests/contrib/rl/test_replay_buffer.py to src/ares/contrib/rl/replay_buffer_test.py. Pytest discovers both locations.

File: src/ares/contrib/rl/replay_buffer.py [hunk 10]
```diff
@@ -0,0 +1,583 @@
...
+    obs_t: ObservationType
+    action_t: ActionType
+    rewards_seq: list[float]
+    next_obs: ObservationType
+    terminal: bool
+    next_discount: float
+    discount_powers: list[float]
+    start_step: int
+    actual_n: int
+    gamma: float
```
Comment (joshgreaves) [inline src/ares/contrib/rl/replay_buffer.py]: Note to self: re-review these and work out what we should include, and what we shouldn't.
Consider ordering in the code.

File: src/ares/contrib/rl/replay_buffer.py [hunk 11]
```diff
@@ -0,0 +1,585 @@
...
+        gamma: float,
+    ) -> NStepSample:
+        """
+        Build an n-step sample starting from a given position.
+
+        Never crosses episode boundary; truncates if fewer than n steps remain.
+        """
+        episode = self._episodes[episode_id]
+
+        num_steps = len(episode.actions)
```
[the remark this answers is not in the document]
Comment (joshgreaves) [reply src/ares/contrib/rl/replay_buffer.py]: This is still not resolved correctly. `num_steps = len(episode.actions)` is not correct, since if you have exactly one action and 1 state you have 0 transitions (since you need an s').

Please change this for `num_steps = len(episode)` and implement `__len__` on episode to be `max(len(states) - 1, 0)`

File: src/ares/contrib/rl/replay_buffer.py [hunk 12]
```diff
@@ -0,0 +1,583 @@
...
+            oldest = min(finished_episodes, key=lambda x: x[1].start_time)
+            episode_id = oldest[0]
+        else:
+            # No finished episodes, evict oldest in-progress
+            oldest = min(in_progress_episodes, key=lambda x: x[1].start_time)
+            episode_id = oldest[0]
+
+        # Remove the episode
+        episode = self._episodes.pop(episode_id)
+        self._total_steps -= len(episode)
```
[the remark this answers is not in the document]
Comment (joshgreaves) [reply src/ares/contrib/rl/replay_buffer.py]: `_total_steps`  should be replaced with `_total_transitions`. Appending data to an Episode should return 0 if it didn't result in a new transition, and 1 if it resulted in a new transition.

File: src/ares/contrib/rl/replay_buffer.py [hunk 13]
```diff
@@ -0,0 +1,585 @@
...
+            #   - t < len(actions)
+            # TODO: For very large buffers, consider using a Fenwick tree
+            # (Binary Indexed Tree) to maintain cumulative step counts per episode,
+            # enabling O(log n) sampling instead of O(num_episodes) scan.
+            valid_positions: list[tuple[str, int]] = []
+
+            for episode_id, episode in self._episodes.items():
+                num_steps = len(episode.actions)
+                if num_steps == 0:
+                    continue
```
[the remark this answers is not in the document]
Comment (rowan-assistant) [reply src/ares/contrib/rl/replay_buffer.py]: Done! I've implemented the deque optimization:

- Added `self._episode_order` deque to track episode IDs in insertion order
- Updated `sample_n_step()` to iterate over the deque instead of `self._episodes.items()`
- Updated `_evict_oldest_episode()` to remove from the deque
- Updated `clear()` to clear the deque
- Episodes with 0 valid steps are handled correctly (skipped in episode_ranges)

The deque maintains insertion order and enables efficient iteration during sampling. All tests pass.

Commit: 06c6c2a

[the remark this answers is not in the document]
Comment (rowan-assistant) [reply src/ares/contrib/rl/replay_buffer.py]: Fixed in commit 93ed816. 

Now maintaining a parallel `_episode_valid_counts` deque that stays synchronized with `_episode_order`. The deque is updated whenever episode valid counts change (on append, end_episode, and eviction). This eliminates the O(num_episodes) scan during sampling while ensuring correct alignment between episode IDs and their valid step counts.
